modin/experimental/pandas/numpy_wrap.py

```diff
@@ -0,0 +1,132 @@
+                import warnings
+
+                warnings.warn(
+                    "Was not able to intercept all numpy imports. "
+                    "To intercept all of these please do 'import modin.experimental.pandas' as early as possible"
+                )
+                self.__has_to_warn = False
+
+        def __update_engine(self, _):
+            if execution_engine.get() == "Cloudray":
+                from modin.experimental.cloud import get_connection
+
+                self.__swap_numpy(get_connection().modules["numpy"])
+            else:
+                self.__swap_numpy()
+
+        def __make_reducer(self, name):
+            try:
+                reducer = self.__reducers[name]
+            except KeyError:
+
+                def reducer(
+                    obj,
+                    real_obj=getattr(real_numpy, name),
+                    real_obj_reducer=getattr(real_numpy, name).__reduce__,
+                ):
+                    # Ref: https://docs.python.org/3.6/library/pickle.html#object.__reduce__
+                    reduced = real_obj_reducer(obj)
+                    if not isinstance(reduced, tuple):
+                        return reduced
+                    assert isinstance(
+                        reduced[0],
+                        (type, types.FunctionType, types.BuiltinFunctionType),
+                    ), "Do not know how to support this reconstructor"
+
+                    modobj = self.__current_numpy
+                    for submod in reduced[0].__module__.split(".")[1:]:
+                        modobj = getattr(modobj, submod)
+                    reconstruct = getattr(modobj, reduced[0].__name__)
+                    # TODO: see if replacing all "real numpy" things in reduced[1:] is needed
+                    return (reconstruct,) + reduced[1:]
+
+                self.__reducers[name] = reducer
+            return reducer
+
+        def __get_numpy(self):
+            frame = sys._getframe()
+            try:
+                caller_file = frame.f_back.f_back.f_code.co_filename
+            except AttributeError:
+                return self.__current_numpy
+            finally:
+                del frame
+            if any(caller_file.startswith(p) for p in _EXCLUDE_PATHS):
+                return real_numpy
+            return self.__current_numpy
+
+        def __getattr__(self, name):
+            obj = getattr(self.__get_numpy(), name)
+            if isinstance(obj, type):
+                copyreg.pickle(obj, self.__make_reducer(name))
+            return obj
+
+        def __setattr__(self, name, value):
+            if name in self.__own_attrs__:
+                super().__setattr__(name, value)
+            else:
+                setattr(self.__get_numpy(), name, value)
+
+        def __delattr__(self, name):
+            if name not in self.__own_attrs__:
+                delattr(self.__get_numpy(), name)
```

Comment:
Please add docstring for these methods. It's not obvious, that in `__setattr__` you use `super()`, but don't do this in `__getattr__` and `__getattr__`.

Reply:
added